arch/xtensa/src/esp32s3/esp32s3_dma.c

```diff
@@ -0,0 +1,380 @@
+ ****************************************************************************/
+
+static bool  g_dma_chan_used[ESP32S3_DMA_CHAN_MAX];
+static sem_t g_dma_exc_sem = SEM_INITIALIZER(1);
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32s3_dma_request
+ *
+ * Description:
+ *   Request DMA channel and config it with given parameters.
+ *
+ * Input Parameters:
+ *   periph   - Peripheral for which the DMA channel request was made
+ *   tx_prio  - Interrupt priority
+ *   rx_prio  - Interrupt flags
+ *   burst_en - Enable burst transmission
+ *
+ * Returned Value:
+ *   DMA channel number (>=0) if success or -1 if fail.
+ *
+ ****************************************************************************/
+
+int32_t esp32s3_dma_request(enum esp32s3_dma_periph_e periph,
+                            uint32_t tx_prio,
+                            uint32_t rx_prio,
+                            bool burst_en)
+{
+  int chan;
+
+  DEBUGASSERT(periph  < ESP32S3_DMA_PERIPH_NUM);
+
+  DEBUGASSERT(tx_prio <= ESP32S3_DMA_TX_PRIO_MAX);
+  DEBUGASSERT(rx_prio <= ESP32S3_DMA_RX_PRIO_MAX);
+
+  dmainfo("periph=%" PRIu32 " tx_prio=%" PRIu32 " rx_prio=%" PRIu32 "\n",
+          (uint32_t)periph, tx_prio, rx_prio);
+
+  nxsem_wait_uninterruptible(&g_dma_exc_sem);
+
+  for (chan = 0; chan < ESP32S3_DMA_CHAN_MAX; chan++)
+    {
+      if (!g_dma_chan_used[chan])
+        {
+          g_dma_chan_used[chan] = true;
+          break;
+        }
+    }
+
+  if (chan == ESP32S3_DMA_CHAN_MAX)
+    {
+      dmaerr("No available GDMA channel for allocation\n");
+
+      nxsem_post(&g_dma_exc_sem);
+
+      return ERROR;
+    }
+
+  dmainfo("Allocated channel=%d\n", chan);
+
+  if (periph == ESP32S3_DMA_PERIPH_MEM)
+    {
+      /* Enable DMA channel M2M mode */
+
+      SET_BITS(DMA_IN_CONF0_CH0_REG, chan, DMA_MEM_TRANS_EN_CH0_M);
+
+      /* Just setting a valid value to the register */
+
+      SET_REG(DMA_OUT_PERI_SEL_CH0_REG, chan, 0);
+      SET_REG(DMA_IN_PERI_SEL_CH0_REG, chan, 0);
+    }
+  else
+    {
+      /* Disable DMA channel M2M mode */
+
+      CLR_BITS(DMA_IN_CONF0_CH0_REG, chan, DMA_MEM_TRANS_EN_CH0_M);
+
+      /* Connect DMA TX/RX channels to a given peripheral */
+
+      SET_REG(DMA_OUT_PERI_SEL_CH0_REG, chan, periph);
+      SET_REG(DMA_IN_PERI_SEL_CH0_REG, chan, periph);
+    }
+
+  if (burst_en)
+    {
+      /* Enable DMA TX/RX channels burst sending data */
+
+      SET_BITS(DMA_OUT_CONF0_CH0_REG, chan, DMA_OUT_DATA_BURST_EN_CH0_M);
+      SET_BITS(DMA_IN_CONF0_CH0_REG, chan, DMA_IN_DATA_BURST_EN_CH0_M);
+
+      /* Enable DMA TX/RX channels burst reading descriptor link */
+
+      SET_BITS(DMA_OUT_CONF0_CH0_REG, chan, DMA_OUTDSCR_BURST_EN_CH0_M);
+      SET_BITS(DMA_IN_CONF0_CH0_REG, chan, DMA_INDSCR_BURST_EN_CH0_M);
```

Comment:
Minor
Maybe change `DMA_OUTDSCR_BURST_EN_CH0_M` and `DMA_INDSCR_BURST_EN_CH0_M` to `DMA_OUT_DSCR_BURST_EN_CH0_M` and `DMA_IN_DSCR_BURST_EN_CH0_M`

Reply:
These registers names and their fields are auto-generated and they are also referenced by the Technical Reference Manual (TRM).
So I'd recommend against this change, it will increase maintenance complexity and will make cross-reference harder with the TRM.

Reply:
Thank you @gustavonihei! @pkarashchenko I removed the name changes and kept only the suggestion to move the "Included Files" to the right position

Reply:
Ok. If those names are from TRM let's keep it. Just was wondering that in some places IN_ is used and in another there is no _